CDTDatastoreTests/CDTSessionCookieInterceptorTests.m

```diff
@@ -324,4 +356,146 @@ - (void)testCookieInterceptorRenewsEarly
+                                               testCookieHeaderValue2]
+                              }];
+        }];
+        
+        // get resource successfully using new cookie
+        [helper addResponse:^OHHTTPStubsResponse *__nonnull(NSURLRequest *__nonnull request) {
+            XCTAssert([testCookieHeaderValue2 isEqualToString: request.allHTTPHeaderFields[@"Cookie"]]);
+            XCTAssert([request.HTTPMethod isEqualToString:@"GET"]);
+            return [OHHTTPStubsResponse responseWithJSONObject:@{@"ok" : @(YES)} statusCode:200 headers:@{}];
+        }];
+    } else {
+        // get resource fails, cookie was not re-newed
+        [helper addResponse:^OHHTTPStubsResponse *__nonnull(NSURLRequest *__nonnull request) {
+            XCTAssert([testCookieHeaderValue isEqualToString: request.allHTTPHeaderFields[@"Cookie"]]);
+            XCTAssert([request.HTTPMethod isEqualToString:@"GET"]);
+            return [OHHTTPStubsResponse responseWithJSONObject:@{@"error": error,
+                                                                 @"reason": reason}
+                                                    statusCode:403 headers:@{}];
+        }];
+    }
+
+    
+    [helper doStubsForHost:@"username403.cloudant.com"];
+    
+    CDTSessionCookieInterceptor *interceptor =
+    [[CDTSessionCookieInterceptor alloc] initWithUsername:@"username403" password:@"password"];
+    // create a context with a request which we can use
+    NSURL *url = [NSURL URLWithString:@"http://username403.cloudant.com/somedb"];
+    NSURLRequest *request = [NSURLRequest requestWithURL:url];
+    
+    CDTURLSession *session = [[CDTURLSession alloc] initWithCallbackThread:[NSThread currentThread] requestInterceptors:@[interceptor] sessionConfigDelegate: nil];
+    
+    // Initial GET will get a cookie and resource
+    CDTURLSessionTask *task = [session dataTaskWithRequest:request taskDelegate:nil];
+    [task resume];
+    while ([task state] != NSURLSessionTaskStateCompleted) {
+        [NSThread sleepForTimeInterval:1.0];
+    }
+    
+    // Second GET should fail with 403 error
+    CDTURLSessionTask *task2 = [session dataTaskWithRequest:request taskDelegate:nil];
+    [task2 resume];
+    while ([task2 state] != NSURLSessionTaskStateCompleted) {
+        [NSThread sleepForTimeInterval:1.0];
+    }
+        
+    if([error isEqualToString:@"credentials_expired"]) {
+        // call to _session endpoint
+        CDTURLSessionTask *task3 = [session dataTaskWithRequest:request taskDelegate:nil];
+        [task3 resume];
+        while ([task3 state] != NSURLSessionTaskStateCompleted) {
+            [NSThread sleepForTimeInterval:1.0];
+        }
+        
+        // 403 'credentials_expired' case: call to _session in task3 re-newed cookie
+        XCTAssertNotNil(interceptor.cookies);
+        
+        // GET resource with new cookie
+        CDTURLSessionTask *task4 = [session dataTaskWithRequest:request taskDelegate:nil];
```

Comment:
I'm confused as to why this 4th task even exists and I think the only reason it isn't causing a failure is because of the extra (uneeded) stub inserted in the setup. I think if you looked at the response code for this task it would be a `403` because of that extra stub. As I wrote in earlier comment I think that setup stub should be removed and this 4th task should be removed.

I think 3 tasks is sufficient to exercise the test path you want in each case. I have summarized the flow I would expect in each case which ought to help with getting the right number of stubs down each path:

 Task | Request/Stub responses (expired) | Request/Stub responses (not-expired)
--- | --- | ---
1 `GET http://username403.cloudant.com/somedb` | `GET _session` -> `200 cookie1` <br/> `GET somedb (with cookie1)` -> `200` | `GET _session` -> `200 cookie1` <br/> `GET somedb (with cookie1)` -> `200`
2 `GET http://username403.cloudant.com/somedb` | `GET somedb (with cookie1)` -> `403 {error: "credentials_expired"}` <br/> `GET _session` -> `200 cookie2` <br/> `GET somedb (with cookie2)` -> `200` | `GET somedb (with cookie1)` -> `403 {error: "403_not_expired_test"}`
3 `GET http://username403.cloudant.com/somedb` | `GET somedb (with cookie2)` -> `200` | `GET somedb (with cookie1)` -> `403 {error: "403_not_expired_test"}`
Total requests made/ stubs needed | 6 | 4

Reply:
@ricellis Thanks for spending the time to write out this summary table.  I've created comments using these details in 565045f.